packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit_webview_widget.dart

```diff
@@ -121,25 +130,42 @@ class WebKitWebViewPlatformController extends WebViewPlatformController {
     didFinishNavigation: (WKWebView webView, String? url) {
       callbacksHandler.onPageFinished(url ?? '');
     },
-  )
-        ..setDidStartProvisionalNavigation((WKWebView webView, String? url) {
-          callbacksHandler.onPageStarted(url ?? '');
-        })
-        ..setDidFailNavigation((WKWebView webView, NSError error) {
-          callbacksHandler.onWebResourceError(_toWebResourceError(error));
-        })
-        ..setDidFailProvisionalNavigation((WKWebView webView, NSError error) {
-          callbacksHandler.onWebResourceError(_toWebResourceError(error));
-        })
-        ..setWebViewWebContentProcessDidTerminate((WKWebView webView) {
-          callbacksHandler.onWebResourceError(WebResourceError(
-            errorCode: WKErrorCode.webContentProcessTerminated,
-            // Value from https://developer.apple.com/documentation/webkit/wkerrordomain?language=objc.
-            domain: 'WKErrorDomain',
-            description: '',
-            errorType: WebResourceErrorType.webContentProcessTerminated,
-          ));
-        });
+    didStartProvisionalNavigation: (WKWebView webView, String? url) {
+      callbacksHandler.onPageStarted(url ?? '');
+    },
+    decidePolicyForNavigationAction: (
+      WKWebView webView,
+      WKNavigationAction action,
+    ) async {
+      if (!_hasNavigationDelegate) {
+        return WKNavigationActionPolicy.allow;
+      }
+
+      final bool allow = await callbacksHandler.onNavigationRequest(
```

Comment:
Why do we need a separate variable tracking whether there's a navigation delegate, vs just null-checking this function?

Reply:
Since [WebViewPlatformCallbacksHandler](https://github.com/flutter/plugins/blob/main/packages/webview_flutter/webview_flutter_platform_interface/lib/src/platform_interface/webview_platform_callbacks_handler.dart#L13) is an abstract class, it should always be nonnull. The platform implementations unfortunately don't have access to the navigation delegate that the user provides. `webview_flutter` handles the null check in the `WebViewPlatformCallbacksHandler` implementaiton. https://github.com/flutter/plugins/blob/main/packages/webview_flutter/webview_flutter/lib/src/webview.dart#L451.